include/deal.II/differentiation/ad/ad_drivers.h

```diff
@@ -0,0 +1,1270 @@
+      const unsigned int &,
+      const std::vector<ScalarType> &,
+      Vector<ScalarType> &)
+    {
+      AssertThrow(false, ExcRequiresADNumberSpecialization());
+    }
+
+
+    template<typename ADNumberType, typename ScalarType, typename T>
+    void
+    TapedDrivers<ADNumberType,ScalarType,T>::hessian (
+      const unsigned int &,
+      const std::vector<ScalarType> &,
+      FullMatrix<ScalarType> &)
+    {
+      AssertThrow(false, ExcRequiresADNumberSpecialization());
+    }
+
+
+    template<typename ADNumberType, typename ScalarType, typename T>
+    void
+    TapedDrivers<ADNumberType,ScalarType,T>::values (
+      const unsigned int &,
+      const unsigned int &,
+      const std::vector<ScalarType> &,
+      Vector<ScalarType> &)
+    {
+      AssertThrow(false, ExcRequiresADNumberSpecialization());
+    }
+
+
+    template<typename ADNumberType, typename ScalarType, typename T>
+    void
+    TapedDrivers<ADNumberType,ScalarType,T>::jacobian (
+      const unsigned int &,
+      const unsigned int &,
+      const std::vector<ScalarType> &,
+      FullMatrix<ScalarType> &)
+    {
+      AssertThrow(false, ExcRequiresADNumberSpecialization());
+    }
+
+
+    // Specialization for taped Adol-C auto-differentiable numbers.
+    //
+    // Note: In the case of Adol-C taped numbers, the associated scalar
+    // type is always expected to be a double. So we need to make a further
+    // specialization when ScalarType is a float.
+    template<typename ADNumberType>
+    struct TapedDrivers<ADNumberType,double,typename std::enable_if<
+      ADNumberTraits<ADNumberType>::type_code == NumberTypes::adolc_taped
+      >::type>
+    {
+      typedef double scalar_type;
+
+      // === Taping ===
+
+      static void
+      enable_taping(const unsigned int &tape_index,
+                    const bool         &keep)
+      {
+        trace_on(tape_index,keep);
+      }
+
+      static void
+      disable_taping(const unsigned int &active_tape_index,
+                     const bool         &write_tapes_to_file)
+      {
+#ifdef DEAL_II_WITH_ADOLC
+        if (write_tapes_to_file)
+          {
+            trace_off(active_tape_index); // Slow
+            std::vector<std::size_t> counts (STAT_SIZE);
+            ::tapestats(active_tape_index, counts.data());
+          }
+        else
+          trace_off(); // Fast(er)
+#else
+        AssertThrow(false, ExcRequiresAdolC());
+#endif
+      }
+
+      template<typename Stream>
+      static void
+      print_tape_stats(Stream             &stream,
+                       const unsigned int &tape_index)
+      {
+        // See Adol-C manual section 2.1
+        // and adolc/taping.h
+        std::vector<std::size_t> counts (STAT_SIZE);
+        ::tapestats(tape_index, counts.data());
+        Assert(counts.size() >= 18, ExcInternalError());
+        stream
+            << "Tape index: " << tape_index << "\n"
+            << "Number of independent variables: " << counts[0] << "\n"
```

Comment:
`std::endl` ?

Reply:
`std::endl` flushes the write buffer, so its a bit wasteful to do this 18 times in succession. I do a final `std::flush` at the end.

Reply:
ah, i see.